File: src/protocol.ts
export interface NativeVoice {
  id: string;
  name: string;
  locale?: string;
  gender?: string;
}

export type PreparedUtteranceKind =
  | "heading"
  | "paragraph"
  | "item"
  | "block_quote"
  | "table_row";

export interface PreparedUtterance {
  utterance_index: number;
  text: string;
  start_offset: number;
  end_offset: number;
  kind: PreparedUtteranceKind;
}

export interface NativeUtteranceEvent {
  type: "utterance_begin" | "utterance_end";
  document_id: string;
  utterance_index: number;
  start_offset: number;
  end_offset: number;
  text: string;
}

export interface NativeBackendUnavailableEvent {
  type: "backend_unavailable";
  message: string;
}

export type NativeEvent = NativeUtteranceEvent | NativeBackendUnavailableEvent;

export interface InitResult {
  id: string;
  type: "init_result";
  available: boolean;
  backend: string;
  message?: string;
}

export interface ListVoicesResult {
  id: string;
  type: "list_voices_result";
  voices: NativeVoice[];
}

export interface LoadDocumentResult {
  id: string;
  type: "load_document_result";
  document_id: string;
  utterance_count: number;
}

export interface PrepareSpeechResult {
  id: string;
  type: "prepare_speech_result";
  document_id: string;
  language_code?: string;
  utterances: PreparedUtterance[];
}

export interface PrepareSummaryResult {
  id: string;
  type: "prepare_summary_result";
  document_id: string;
  language_code?: string;
  utterances: PreparedUtterance[];
}

export interface SpeakResult {
  id: string;
  type: "speak_result";
  document_id: string;
  queued: number;
}

export interface StopResult {
  id: string;
  type: "stop_result";
}

export interface ErrorResult {
  id: string;
  type: "error_result";
  code: string;
  message: string;
}

export type NativeResponse =
  | InitResult
  | ListVoicesResult
  | LoadDocumentResult
  | PrepareSpeechResult
  | PrepareSummaryResult
  | SpeakResult
  | StopResult
  | ErrorResult;

export interface InitRequest {
  id: string;
  type: "init";
}

export interface ListVoicesRequest {
  id: string;
  type: "list_voices";
}

export interface LoadDocumentRequest {
  id: string;
  type: "load_document";
  document_id: string;
  text: string;
}

export interface PrepareSpeechRequest {
  id: string;
  type: "prepare_speech";
  document_id: string;
  start_offset?: number;
  end_offset?: number;
}

export interface PrepareSummaryRequest {
  id: string;
  type: "prepare_summary";
  document_id: string;
}

export interface SpeakRequest {
  id: string;
  type: "speak";
  document_id: string;
  start_offset?: number;
  end_offset?: number;
  voice_id?: string;
  rate: number;
}

export interface SpeakSummaryRequest {
  id: string;
  type: "speak_summary";
  document_id: string;
  voice_id?: string;
  rate: number;
}

export interface StopRequest {
  id: string;
  type: "stop";
}

export type NativeRequest =
  | InitRequest
  | ListVoicesRequest
  | LoadDocumentRequest
  | PrepareSpeechRequest
  | PrepareSummaryRequest
  | SpeakRequest
  | SpeakSummaryRequest
  | StopRequest;

File: native/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Request {
    Init {
        id: String,
    },
    ListVoices {
        id: String,
    },
    LoadDocument {
        id: String,
        document_id: String,
        text: String,
    },
    PrepareSpeech {
        id: String,
        document_id: String,
        start_offset: Option<usize>,
        end_offset: Option<usize>,
    },
    PrepareSummary {
        id: String,
        document_id: String,
    },
    Speak {
        id: String,
        document_id: String,
        start_offset: Option<usize>,
        end_offset: Option<usize>,
        voice_id: Option<String>,
        rate: f32,
    },
    SpeakSummary {
        id: String,
        document_id: String,
        voice_id: Option<String>,
        rate: f32,
    },
    Stop {
        id: String,
    },
}

#[derive(Clone, Debug, Serialize)]
pub struct VoiceInfo {
    pub id: String,
    pub name: String,
    pub locale: Option<String>,
    pub gender: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum PreparedUtteranceKind {
    Heading,
    Paragraph,
    Item,
    BlockQuote,
    TableRow,
}

#[derive(Clone, Debug, Serialize)]
pub struct PreparedUtterance {
    pub utterance_index: usize,
    pub text: String,
    pub start_offset: usize,
    pub end_offset: usize,
    pub kind: PreparedUtteranceKind,
}

#[derive(Clone, Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Message {
    InitResult {
        id: String,
        available: bool,
        backend: String,
        message: Option<String>,
    },
    ListVoicesResult {
        id: String,
        voices: Vec<VoiceInfo>,
    },
    LoadDocumentResult {
        id: String,
        document_id: String,
        utterance_count: usize,
    },
    PrepareSpeechResult {
        id: String,
        document_id: String,
        language_code: Option<String>,
        utterances: Vec<PreparedUtterance>,
    },
    PrepareSummaryResult {
        id: String,
        document_id: String,
        language_code: Option<String>,
        utterances: Vec<PreparedUtterance>,
    },
    SpeakResult {
        id: String,
        document_id: String,
        queued: usize,
    },
    StopResult {
        id: String,
    },
    ErrorResult {
        id: String,
        code: String,
        message: String,
    },
    UtteranceBegin {
        document_id: String,
        utterance_index: usize,
        start_offset: usize,
        end_offset: usize,
        text: String,
    },
    UtteranceEnd {
        document_id: String,
        utterance_index: usize,
        start_offset: usize,
        end_offset: usize,
        text: String,
    },
    BackendUnavailable {
        message: String,
    },
}
